### DISCUSSION TITLE
Valid constraints for non-binary references?

### DISCUSSION THREAD

**Discussion by tylerandrewscott** [statnet/ergm#295] [Q&A] [answered]
Hello, I'm having a little trouble figuring out how the constraints=~ options work/do not work in the context of, say, a Unif(0,1) reference ergm. Ideally I would like to constrain for the density of nonzero dyads or fix the degree of only a few high nonzero-degree nodes, but those sorts of options don't seem to work (or perhaps I'm just missing something obvious?).

**Comment by krivit** [comment, marked as answer]
Briefly, the way estimation works is that `ergm` takes the combination of constraints you've given it, plus the reference distribution and a few other things, and tries to figure out if it has a Metropolis-Hastings proposal that can sample from a space of networks that satisfies those constraints. If it doesn't, it stops with an error. At the moment, there are relatively few valued ERGM proposals that properly enforce constraints in `ergm`.

More generally, I don't know whether `Unif(0,1)` reference makes sense if you are going to have dyads that are 0 but are not fixed at 0. This is because for a continuous random variable *X*~Unif(0,1), Pr(*X* = 0) = 0 (and, more generally Pr(*X* = *x*) = 0 for any constant *x*). Your problem therefore probably requires a custom Metropolis-Hastings proposal.

**Comment by tylerandrewscott** [comment]
Ah, that makes total sense and is super helpful, thanks! -tyler

On Sun, May 16, 2021 at 1:21 AM Pavel N. Krivitsky ***@***.***>
wrote:

> Briefly, the way estimation works is that ergm takes the combination of
> constraints you've given it, plus the reference distribution and a few
> other things, and tries to figure out if it has a Metropolis-Hastings
> proposal that can sample from a space of networks that satisfies those
> constraints. If it doesn't, it stops with an error. At the moment, there
> are relatively few valued ERGM proposals that properly enforce constraints
> in ergm.
>
> More generally, I don't know whether Unif(0,1) reference makes sense if
> you are going to have dyads that are 0 but are not fixed at 0. This is
> because for a continuous random variable *X*~Unif(0,1), Pr(*X* = 0) = 0
> (and, more generally Pr(*X* = *x*) = 0 for any constant *x*). Your
> problem therefore probably requires a custom Metropolis-Hastings proposal.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/statnet/ergm/discussions/295#discussioncomment-743789>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4G7X4ANBTTCPFYHEPHNJLTN557LANCNFSM446WQ3CA>
> .
>


-- 
Tyler Scott
***@***.***
Department of Environmental Science and Policy
University of California, Davis
